crates/matrix-sdk-ui/src/authentication/oidc.rs

```diff
@@ -0,0 +1,214 @@
+        let oidc_directory = PathBuf::from(base_path).join("oidc");
+        fs::create_dir_all(&oidc_directory).map_err(|_| OidcRegistrationsError::InvalidBasePath)?;
+
+        let metadata_hash = {
+            let mut hasher = DefaultHasher::new();
+            let metadata_string = serde_json::to_string(metadata).unwrap();
+            metadata_string.hash(&mut hasher);
+            hasher.finish()
+        };
+
+        Ok(OidcRegistrations {
+            file_path: oidc_directory.join("registrations.json"),
+            metadata_hash,
+            static_registrations,
+        })
+    }
+
+    /// Returns the underlying registration data.
+    fn registration_data(&self) -> RegistrationData {
+        let reader = match File::open(&self.file_path) {
+            Ok(file) => BufReader::new(file),
+            Err(e) => {
+                tracing::warn!("Failed to open registrations file: {e}");
+                return RegistrationData {
+                    metadata_hash: self.metadata_hash,
+                    dynamic_registrations: Default::default(),
+                };
+            }
+        };
+
+        let registration_data = match serde_json::from_reader::<_, RegistrationData>(reader) {
+            Ok(data) => data,
+            Err(e) => {
+                tracing::error!("Failed to parse registrations file: {e}");
+                return RegistrationData {
+                    metadata_hash: self.metadata_hash,
+                    dynamic_registrations: Default::default(),
+                };
+            }
+        };
+
+        if registration_data.metadata_hash != self.metadata_hash {
+            tracing::warn!("Metadata hash mismatch, clearing registrations");
+            return RegistrationData {
+                metadata_hash: self.metadata_hash,
+                dynamic_registrations: Default::default(),
+            };
+        }
+
+        registration_data
+    }
+
+    /// Returns the client ID registered for a particular issuer or None if a
+    /// registration hasn't been made.
+    pub fn client_id(&self, issuer: &str) -> Option<String> {
+        let mut data = self.registration_data();
+        data.dynamic_registrations.extend(self.static_registrations.clone());
+        data.dynamic_registrations.get(issuer).cloned()
+    }
+
+    /// Stores a new client ID registration for a particular issuer. If a client
+    /// ID has already been stored, this will overwrite the old value.
+    pub fn set_client_id(
+        &self,
+        client_id: String,
+        issuer: String,
+    ) -> Result<(), OidcRegistrationsError> {
+        let mut data = self.registration_data();
+        data.dynamic_registrations.insert(issuer, client_id);
+
+        let writer = BufWriter::new(
+            File::create(&self.file_path).map_err(|_| OidcRegistrationsError::InvalidBasePath)?,
+        );
+        serde_json::to_writer(writer, &data).map_err(|_| OidcRegistrationsError::InvalidBasePath)
```

Comment:
Heh, it's a serialization error, not an invalid base path :P

Reply:
Oh, so it couldn't be a an error writing to the file?

Reply:
It could be that the serialization failed, OR that the writer failed; in fact it might be nice to not ignore the error in `map_err` and report it to the caller (feel free to stringify it before doing that).